### PR TITLE
feat: switch Program Factory to use proxy 

### PR DESCRIPTION
**Pull request by thelostone-mc**
#### Description

- uses proxy pattern on Program to make an upgradeable contract
- introduces dummy upgradable contract with documentation on how to go about this 

This PR shows how we can rely on Proxy for program factory.
Once we focus on test coverage -> we can consider implementing the proxy pattern for Grant Factory as well 

Further details on what the proxy does can be found in the readme file.
For now **this does not change the behaviour on how program are created / deployed**

### CODE REVIEW

File: packages/contracts/scripts/program/deployProgramImplementation.ts [hunk 1]
```diff
@@ -15,10 +15,10 @@ export async function main() {
   });
 
   // Deploy ProgramImplementation 
-  const contractFactory = await ethers.getContractFactory("ProgramImplementation");
+  const contractFactory = await ethers.getContractFactory("DummyProgramImplementation");
```
Comment (chibie) [inline packages/contracts/scripts/program/deployProgramImplementation.ts:18]: why was this renamed to `DummyProgramImplementation` ?

**Review by chibie — COMMENTED** [review]
lgtm... just one question

File: packages/contracts/scripts/program/deployProgramImplementation.ts (hunk 1 above)
Comment (thelostone-mc) [reply packages/contracts/scripts/program/deployProgramImplementation.ts:18]:
```suggestion
  const contractFactory = await ethers.getContractFactory("ProgramImplementation");
```

**Comment by thelostone-mc** [comment]
Ah nice catch ! Forgot to update that back
